lib/std/fmt.zig

```diff
@@ -709,6 +709,87 @@ fn formatFloatValue(
     return formatBuf(buf_stream.getWritten(), options, writer);
 }
 
+fn formatSliceHexImpl(comptime uppercase: bool) type {
+    const charset = "0123456789" ++ if (uppercase) "ABCDEF" else "abcdef";
+
+    return struct {
+        pub fn f(
+            bytes: []const u8,
+            comptime fmt: []const u8,
+            options: std.fmt.FormatOptions,
+            writer: anytype,
+        ) !void {
+            var buf: [2]u8 = undefined;
+
+            for (bytes) |c| {
+                buf[0] = charset[c >> 4];
+                buf[1] = charset[c & 15];
+                try writer.writeAll(&buf);
+            }
+        }
+    };
+}
+
+const formatSliceHexLower = formatSliceHexImpl(false).f;
+const formatSliceHexUpper = formatSliceHexImpl(true).f;
+
+/// Return a Formatter for a []const u8 where every byte is formatted as a pair
+/// of lowercase hexadecimal digits.
+pub fn fmtSliceHexLower(bytes: []const u8) std.fmt.Formatter(formatSliceHexLower) {
+    return .{ .data = bytes };
+}
+
+/// Return a Formatter for a []const u8 where every byte is formatted as a pair
+/// of uppercase hexadecimal digits.
+pub fn fmtSliceHexUpper(bytes: []const u8) std.fmt.Formatter(formatSliceHexUpper) {
```

Comment:
I don't love that we have different functions for this for upper vs lower case.
Could we instead take the case as an argument?

`pub fn fmtSliceHex(bytes: []const u8, case: enum{Upper, Lower}) std.fmt.Formatter(formatSliceHex) {`

Reply:
No strong opinion on having a single formatter, I went with two separate ones to keep it close to the `x`\`X`.

(IIRC there's a stage1 bug regarding comptime anonymous enum arguments, check out debug.zig)